File: Skills Demo 2/Python/craps.py
# NOTE Game logic/rules
# When player starts game, they need to be given a point
#   - Points are based on rolling 2 dice (2-12)
# LOSE:
#   - First roll is 2, 3 or 12
# WIN:
#   - First roll is 7 or 11
# If you get neither, that becomes your point
#   - Roll again to get new point.
# If the new sum is equal your original point, you win
# If the new sum is equal to 7, you lose
# If you get neither roll again
import random


# Roll dice and give user a number between 2 and 12
diceRoll = random.randint(2, 12)


def play():
    # Check for a win
    if diceRoll == 7 or diceRoll == 11:
        print(f'Point: {diceRoll}    Result: WIN')
        print("===================")
        # TODO Add a replay option (if else with while loop to play())
    # Check for a loss
    elif diceRoll == 2 or diceRoll == 3 or diceRoll == 12:
        print(f'Point: {diceRoll}    Result: LOST')
        print("===================")
        # TODO Add a replay option (if else with while loop to play())
    # Neither win or loss
    else:
        print(f'Point: {diceRoll}    Result: Roll again!')
        # Roll another point for the user to check against their original score
        newDiceRoll = random.randint(2, 12)
        while (diceRoll != newDiceRoll or diceRoll == newDiceRoll):
            # Check for a win
            if diceRoll == newDiceRoll:
                # Win
                print(f'Point: {newDiceRoll}    Result: WIN')
                print("===================")
                break
            elif newDiceRoll == 7:
                # Lost
                print(f'Point: {newDiceRoll}    Result: LOST')
                print("===================")
                break
            else:
                # No win or loss
                print(f'Point: {newDiceRoll}    Result: Roll again!')
            newDiceRoll = random.randint(2, 12)


play()

File: Skills Demo 2/Java/Craps.java
package skillsdemotwocrapsgameclass;
import java.util.Random;
import java.util.Scanner;

public class Craps {

 public static void main(String[] args) {
  Scanner userInput = new Scanner(System.in);
  String ans = ""; //Decalre variable for use in play again loop
  boolean game = true;
  while (game == true) {
   RandomNumber diceRoll = new RandomNumber(); //Create instance diceRoll of class RandomNumber
   play playGame = new play(); //Create instance playGame of class play

   //Initialise variables from play class
   playGame.diceRoll = diceRoll.randNum;
   playGame.point = playGame.diceRoll;
   playGame.newPoint = diceRoll.randNum;


   System.out.println("\nWELCOME TO A GAME OF CRAPS!");

   if (playGame.diceRoll == 7 || playGame.diceRoll == 11) {
    //Show the users point
    System.out.println("Point: " + playGame.point);
    System.out.println("-----------------------------------------");

    //Tell user they won
    System.out.println("Congratulations you won, with a " + playGame.diceRoll);
    System.out.println("\nWould you like to play again? [Y/N]\n");
    ans = userInput.next();
    if (ans.equals("n") || ans.equals("N")) {
     game = false;
    } else if (ans.equals("y") || ans.equals("Y")) {
     game = true;
    } else {
     System.out.println("Error! PLease enter either Y or N.");
     System.out.println("\nWould you like to play again? [Y/N]\n");
     ans = userInput.nextLine();
    }

   } else if (playGame.diceRoll == 2 || playGame.diceRoll == 3 || playGame.diceRoll == 12) {
    //Show the users point
    System.out.println("Point: " + playGame.point);
    System.out.println("-----------------------------------------");

    //Tell the user they lost
    System.out.println("Sorry you lost, with a " + playGame.diceRoll);
    System.out.println("\nWould you like to play again? [Y/N]\n");
    ans = userInput.next();
    if (ans.equals("n") || ans.equals("N")) {
     game = false;
    } else if (ans.equals("y") || ans.equals("Y")) {
     game = true;
    } else {
     System.out.println("Error! PLease enter either Y or N.");
     System.out.println("\nWould you like to play again? [Y/N]\n");
     ans = userInput.nextLine();
    }

   } else {
    System.out.println("Point: " + playGame.point);
    System.out.println("-----------------------------------------");

    while(playGame.point != playGame.newPoint || playGame.point == playGame.newPoint) {
     /*
     BUG: (2/2/19)
         User will receive their original roll again, causing them to always win

     FIX: (27/2/19)
         Added method to RandomNumber class, to generate a new number. Then call it before the next move
     */

     //Roll dice again for the new point
     diceRoll.Roll(); //Gets new random number
     playGame.newPoint = diceRoll.randNum; //Sets the newPoint variable, to be given value of the new random number

     //Checks if the user can win
     if (playGame.point == playGame.newPoint) {
      System.out.println("Your new roll: " + playGame.newPoint + "\t\t Win");

      System.out.println("\nWould you like to play again? [Y/N]\n");
      ans = userInput.next();
      if (ans.equals("n") || ans.equals("N")) {
       game = false;
      } else if (ans.equals("y") || ans.equals("Y")) {
       game = true;
      } else {
       System.out.println("Error! PLease enter either Y or N.");
       System.out.println("\nWould you like to play again? [Y/N]\n");
       ans = userInput.nextLine();
      }
      break;
     }
     //Checks if the user has lost
     else if (playGame.newPoint == 7) {
      System.out.println("Your new roll: " + playGame.newPoint + "\t\t Lose");

      /*
      BUG: (27/2/19)
          The while loop would not terminate even if user, did not want to play again.
          This occured everytime the user won or lost, and asked if they wanted to play again.
      FIX: (27/2/19)
          Changed condition in if statement to be .equals instead of ==
      */
      System.out.println("\nWould you like to play again? [Y/N]\n");
      ans = userInput.next();
      if (ans.equals("n") || ans.equals("N")) {
       game = false;
      } else if (ans.equals("y") || ans.equals("Y")) {
       game = true;
      } else {
       System.out.println("Error! PLease enter either Y or N.");
       System.out.println("\nWould you like to play again? [Y/N]\n");
       ans = userInput.nextLine();
      }
      break;
     }
     //Check if user needs to roll again
     else {
      System.out.println("Your new roll: " + playGame.newPoint + "\t\t No help");
     }

    }
   }
  }
 }
}

class RandomNumber {
 Random rand = new Random();
 int randNum = rand.nextInt(12) + 1;
 void Roll() {
  randNum = rand.nextInt(12) + 1;
 }

}

class play {
 int diceRoll, point, newPoint;
}